auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -587,6 +519,120 @@ public MethodSpec createReadMethod(ClassName className,
+
+    ClassName builderClassName = ClassName.get(builderContext.builderClass());
+    FieldSpec builder = FieldSpec.builder(builderClassName, "_builder").build();
+    if (defaultBuilder.isPresent()) {
+      ClassName baseClass = ClassName.get(autoValueClassElement);
+      readMethod.addStatement("$T $N = $T.$L", builder.type, builder, baseClass, defaultBuilder.get());
+    } else {
+      ClassName finalBuilderClassName = className.nestedClass(builderClassName.simpleName());
+      readMethod.addStatement("$T $N = new $T()", builder.type, builder, finalBuilderClassName);
+    }
+
+    readMethod.addStatement("$N.beginObject()", jsonReader);
+
+    readMethod.beginControlFlow("while ($N.hasNext())", jsonReader);
+
+    FieldSpec name = FieldSpec.builder(String.class, "_name").build();
+    readMethod.addStatement("$T $N = $N.nextName()", name.type, name, jsonReader);
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("continue");
+    readMethod.endControlFlow();
+
+    readMethod.beginControlFlow("switch ($N)", name);
+    for (Property prop : properties) {
+      if (!prop.shouldDeserialize()) {
+        continue;
+      }
+
+      for (String alternate : prop.serializedNameAlternate()) {
+        readMethod.addCode("case $S:\n", alternate);
+      }
+      readMethod.beginControlFlow("case $S:", prop.serializedName());
+      Set<ExecutableElement> setters = builderContext.setters().get(prop.humanName);
+
+      // search overloaded methods for a type match
+      Optional<CodeBlock> setterBlock = setters.stream()
+          .filter(e -> ClassName.get(e.getParameters().get(0).asType()).equals(prop.type))
+          .findFirst()
+          .map(method ->
+              CodeBlock.builder()
+                  .addStatement("_builder.$N($N.read($N))", method.getSimpleName(), adapters.get(prop), jsonReader)
+                  .build());
+
+      // if the property is Optional<T>, see if there is an overloaded setter that accepts T
+      if (!setterBlock.isPresent() && prop.instanceOfOptional) {
+        final ClassName optionalPropTypeArgument = (ClassName) ((ParameterizedTypeName) prop.type).typeArguments.get(0);
+        setterBlock = setters.stream()
+            .filter(e -> ClassName.get(e.getParameters().get(0).asType()).equals(optionalPropTypeArgument))
+            .findFirst()
+            .map(method ->
+            {
+              String propertyVarName = prop.humanName + "Property";
+              return CodeBlock.builder()
+                  .addStatement("$T $L = $N.read($N)", prop.type, propertyVarName,
+                      adapters.get(prop), jsonReader)
+                  .beginControlFlow("if ($L.isPresent())", propertyVarName)
+                  .addStatement("_builder.$N($L.get())", method.getSimpleName(), propertyVarName)
```

Comment:
same here with using the `builder` param and making the reference in the statement use `$N`